### PR TITLE
test(cluster): Test that cluster works with a standard cluster client.

### PR DESCRIPTION
**Pull request by chakaz**
In this case, `redis.RedisCluster`.

To be double sure I also looked at the actual packets and saw that the client asks for `CLUSTER SLOTS`, and then after the redistribution of slots, following a few `MOVED` replied, it asks for the new slots again.

### CODE REVIEW

File: tests/dragonfly/cluster_test.py [hunk 1]
```diff
@@ -284,3 +284,125 @@ async def test_cluster_slot_ownership_changes(df_local_factory):
...
+      ]
+    """
+    await push_config(config, c_nodes_admin)
+
+    client = redis.RedisCluster(decode_responses=True, host="localhost", port=nodes[0].port)
+
+    for i in range(10_000):
+        key = 'key' + str(i)
+        assert client.set(key, 'value') == True
+        assert client.get(key) == 'value'
```
Comment (dranikpg) [inline tests/dragonfly/cluster_test.py:354]: Small note: you're using the sync client inside an async function

Comment (chakaz) [reply tests/dragonfly/cluster_test.py:354]: Thanks, that's a good catch!
I'm new to Python async (and also to Python in general).
My goal here is to use `RedisCluster`, which seems to be a sync library. Is there a way for me to still use the other async tools we have here, like `push_config()` and `get_node_id()`?
Alternatively I guess I could use `run_in_executor()` (from [here](https://stackoverflow.com/a/54685298)), do you think that's advisable?

Comment (dranikpg) [reply tests/dragonfly/cluster_test.py:354]: You can use the async `aioredis.RedisCluster` (which is now just a re-export of the official redis.asyncio). It should have the same interface

If there were only the sync version, wouldn't have caused any trouble either way, as you only perform sequential operations and don't have any tasks running

Comment (chakaz) [reply tests/dragonfly/cluster_test.py]: That worked, thanks!

**Comment by adiholden** [comment]
@chakaz This looks good. Will you please add also replica so we will check our client commands with replicas

File: tests/dragonfly/cluster_test.py [hunk 2]
```diff
@@ -396,3 +396,176 @@ async def test_cluster_replica_sets_non_owned_keys(df_local_factory):
...
+          ]
+        }}
+      ]
+    """
+    await push_config(config, c_masters_admin + c_replicas_admin)
+
+    for i in range(10_000):
+        key = 'key' + str(i)
+        assert await client.set(key, 'value') == True
+        assert await client.get(key) == 'value'
```
Comment (dranikpg) [inline tests/dragonfly/cluster_test.py:570]: About the recent timeout in CI that we had. In the future:
- such parts can be split in multiple tasks that run concurrently, all on its range or modulo
- possibly ranges that surely belong to a single node can be handled in pipelines?

When more tests are added, it can add up to hundreds of thousands or requests (its already 40k for this test). Github runners in debug mode have a very low throughput.

**Review by dranikpg — COMMENTED** [review]
Python part looks valid, left a small performance comment. You could've formatted the config in a more compact way to save some lines and make it easier to read, but it probably doesn't matter in this test. 

I don't track the cluster progress in depth, so please don't rely on me in reviews

File: tests/dragonfly/cluster_test.py (hunk 2 above)
Comment (dranikpg) [reply tests/dragonfly/cluster_test.py:570]: Whats also under question if this load is needed. Replication tests trigger all kinds of edge cases with different values and higher load (replicate during update, etc...). Here it seems like it doesn't really matter and you verify only correctness. So you could for example check 10% random values of the range: if you have a bug it's likely to be found in one or a few passes, otherwise why check the whole range

Comment (chakaz) [reply tests/dragonfly/cluster_test.py]: I did not realize we're so constraint!
Thanks for the comment, I tuned down the test.

**Review by dranikpg — APPROVED** [review]

File: tests/dragonfly/cluster_test.py [hunk 3]
```diff
@@ -396,3 +396,176 @@ async def test_cluster_replica_sets_non_owned_keys(df_local_factory):
...
+
+    client = aioredis.RedisCluster(decode_responses=True, host="localhost", port=masters[0].port)
+
+    for i in range(100):
+        key = 'key' + str(i)
+        assert await client.set(key, 'value') == True
+        assert await client.get(key) == 'value'
+
+    # Make sure that getting a value from a replica works as well.
+    replica_response = await client.execute_command(
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:496]: you dont wait here for the replica to finish replicating all the changes

File: tests/dragonfly/cluster_test.py [hunk 4]
```diff
@@ -396,3 +396,176 @@ async def test_cluster_replica_sets_non_owned_keys(df_local_factory):
...
+                "port": {replicas[2].port}
+              }}
+          ]
+        }}
+      ]
+    """
+    await push_config(config, c_masters_admin + c_replicas_admin)
+
+    for i in range(100):
+        key = 'key' + str(i)
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:568]: I think that if you want to set only 100 keys you should generate a random key so that we test all slots when this test runs few times

File: tests/dragonfly/cluster_test.py [hunk 5]
```diff
@@ -396,3 +396,178 @@ async def test_cluster_replica_sets_non_owned_keys(df_local_factory):
...
+          ]
+        }}
+      ]
+    """
+    await push_config(config, c_masters_admin + c_replicas_admin)
+
+    client = aioredis.RedisCluster(decode_responses=True, host="localhost", port=masters[0].port)
+
+    for i in range(100):
+        key = 'key' + str(i)
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:491]: also here random key.
Maybe create nested function , you have the same loop below

File: tests/dragonfly/cluster_test.py [hunk 6]
```diff
@@ -487,11 +487,13 @@ async def test_cluster_native_client(df_local_factory):
 
     client = aioredis.RedisCluster(decode_responses=True, host="localhost", port=masters[0].port)
 
-    for i in range(100):
-        key = 'key' + str(i)
-        assert await client.set(key, 'value') == True
-        assert await client.get(key) == 'value'
+    async def test_random_keys():
+        for i in range(100):
+            key = 'key' + str(i)
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:492]: you forgot to add the random call

Comment (chakaz) [reply tests/dragonfly/cluster_test.py]: whoops, nice catch :)
